pkg/process/monitor/process_monitor.go

```diff
@@ -119,282 +100,239 @@ type ProcessCallback struct {
-					}
-				case *netlink.ExitProcEvent:
-					for _, c := range pm.procEventCallbacks[EXIT] {
-						pm.exitCount += 1
-						pm.evalEXITCallback(c, ev.ProcessPid)
-					}
-					delete(pm.runningPids, ev.ProcessPid)
-				}
-				pm.m.Unlock()
+			pm.eventCount.Inc()
 
-			case err, ok := <-pm.errors:
-				if !ok {
-					return
-				}
-				log.Errorf("process monitor error: %s", err)
-				pm.terminateProcessMonitor()
-				terminated = true
-				pm.Stop()
+			switch ev := event.Msg.(type) {
+			case *netlink.ExecProcEvent:
+				pm.execCount.Inc()
+				pm.handleProcessExec(int(ev.ProcessPid))
+			case *netlink.ExitProcEvent:
+				pm.exitCount.Inc()
+				pm.handleProcessExit(int(ev.ProcessPid))
+			}
+		case err, ok := <-pm.netlinkErrorsChannel:
+			if !ok {
 				return
-
-			case <-logTicker.C:
-				pm.logStats()
 			}
+			log.Errorf("process monitor error: %s", err)
+			log.Info("re-initializing process monitor")
+			pm.netlinkDoneChannel <- struct{}{}
+			// Netlink might suffer from temporary errors (insufficient buffer for example). We're trying to recover
+			// by reinitializing netlink socket.
+			if err := pm.initNetlinkProcessEventMonitor(); err != nil {
+				log.Errorf("failed re-initializing process monitor: %s", err)
+				return
+			}
+		case <-logTicker.C:
+			log.Debugf("process monitor stats - total events: %d; exec events: %d; exit events: %d; Channel size: %d",
+				pm.eventCount.Swap(0), pm.execCount.Swap(0), pm.exitCount.Swap(0), len(pm.netlinkEventsChannel))
 		}
-	}()
-
-	fn := func(pid int) error {
-		for _, c := range pm.procEventCallbacks[EXEC] {
-			pm.evalEXECCallback(c, uint32(pid))
-		}
-		return nil
 	}
+}
 
-	if err := util.WithAllProcs(util.HostProc(), fn); err != nil {
-		return fmt.Errorf("process monitor init, scanning all process failed %s", err)
-	}
-	// enable events to be processed
-	pm.isInitialized = true
-	return nil
+// Initialize setting up the process monitor only once, no matter how many times it was called.
+// The initialization order:
+//  1. Initializes callback workers.
+//  2. Initializes the netlink process monitor.
+//  2. Run the main event loop in a goroutine.
+//  4. Scans already running processes and call the Exec callbacks on them.
+func (pm *ProcessMonitor) Initialize() error {
+	processMonitor.refcount.Inc()
+	var initErr error
+	pm.initOnce.Do(
+		func() {
+			pm.done = make(chan struct{})
+			pm.initCallbackRunner()
+
+			pm.processMonitorWG.Add(1)
+			// Setting up the main loop
+			pm.netlinkDoneChannel = make(chan struct{})
+			pm.netlinkErrorsChannel = make(chan error, 10)
+			pm.netlinkEventsChannel = make(chan netlink.ProcEvent, processMonitorEventQueueSize)
+
+			go pm.mainEventLoop()
+
+			if err := util.WithRootNS(util.GetProcRoot(), func() error {
+				return netlink.ProcEventMonitor(pm.netlinkEventsChannel, pm.netlinkDoneChannel, pm.netlinkErrorsChannel)
+			}); err != nil {
+				initErr = fmt.Errorf("couldn't initialize process monitor: %w", err)
```

Comment:
[nit] Can't we call `initNetlinkProcessEventMonitor` here to avoid these duplicated initialisation code?

Reply:
unfortunately no, as initNetlinkProcessEventMonitor reinitialize the channels, and we want them not to be null when spawning `mainEventLoop`